// ra-input-markdown/lib/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var react_mde_1 = __importDefault(require("react-mde"));
var Showdown = __importStar(require("showdown"));
var ra_core_1 = require("ra-core");
var prop_types_1 = __importDefault(require("prop-types"));
var FormControl_1 = __importDefault(require("@material-ui/core/FormControl"));
var styles_1 = require("@material-ui/core/styles");
require("react-mde/lib/styles/css/react-mde-all.css");
var styles = {};
var MarkdownInput = /** @class */ (function (_super) {
    __extends(MarkdownInput, _super);
    function MarkdownInput(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            value: '',
            tab: 'write'
        };
        return _this;
    }
    MarkdownInput.prototype.componentDidMount = function () {
        var value = this.props.input.value;
        this.setState({ value: value });
        this.converter = new Showdown.Converter({
            tables: true,
            simplifiedAutoLink: true,
            strikethrough: true,
            tasklists: true
        });
    };
    MarkdownInput.prototype.handleValueChange = function (value) {
        this.setState({ value: value });
        this.props.input.onChange(value);
    };
    ;
    MarkdownInput.prototype.handleTabChange = function (tab) {
        this.setState({ tab: tab });
    };
    ;
    MarkdownInput.prototype.render = function () {
        var _this = this;
        return (react_1.default.createElement(FormControl_1.default, { fullWidth: true, className: 'ra-input-mde' },
            react_1.default.createElement(react_mde_1.default, { onChange: function (value) { return _this.handleValueChange(value); }, onTabChange: function (tab) { return _this.handleTabChange(tab); }, value: this.state.value, generateMarkdownPreview: function (markdown) { return Promise.resolve(_this.converter.makeHtml(markdown)); }, selectedTab: this.state.tab })));
    };
    MarkdownInput.propTypes = {
        input: prop_types_1.default.object,
        source: prop_types_1.default.string,
    };
    return MarkdownInput;
}(react_1.Component));
var MarkDownInputWithField = ra_core_1.addField(styles_1.withStyles(styles)(MarkdownInput));
MarkDownInputWithField.defaultProps = {
    addLabel: true,
    fullWidth: true,
};
exports.default = MarkDownInputWithField;
