// ra-ui-materialui/lib/detail/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TabbedShowLayoutTabs = exports.Tab = exports.TabbedShowLayout = exports.SimpleShowLayout = exports.ShowGuesser = exports.ShowActions = exports.ShowView = exports.Show = exports.EditGuesser = exports.EditActions = exports.EditView = exports.Edit = exports.CreateActions = exports.CreateView = exports.Create = void 0;
var Create_1 = __importStar(require("./Create"));
exports.Create = Create_1.default;
Object.defineProperty(exports, "CreateView", { enumerable: true, get: function () { return Create_1.CreateView; } });
var CreateActions_1 = __importDefault(require("./CreateActions"));
exports.CreateActions = CreateActions_1.default;
var Edit_1 = __importStar(require("./Edit"));
exports.Edit = Edit_1.default;
Object.defineProperty(exports, "EditView", { enumerable: true, get: function () { return Edit_1.EditView; } });
var EditActions_1 = __importDefault(require("./EditActions"));
exports.EditActions = EditActions_1.default;
var EditGuesser_1 = __importDefault(require("./EditGuesser"));
exports.EditGuesser = EditGuesser_1.default;
var Show_1 = __importStar(require("./Show"));
exports.Show = Show_1.default;
Object.defineProperty(exports, "ShowView", { enumerable: true, get: function () { return Show_1.ShowView; } });
var ShowActions_1 = __importDefault(require("./ShowActions"));
exports.ShowActions = ShowActions_1.default;
var ShowGuesser_1 = __importDefault(require("./ShowGuesser"));
exports.ShowGuesser = ShowGuesser_1.default;
var SimpleShowLayout_1 = __importDefault(require("./SimpleShowLayout"));
exports.SimpleShowLayout = SimpleShowLayout_1.default;
var TabbedShowLayout_1 = __importDefault(require("./TabbedShowLayout"));
exports.TabbedShowLayout = TabbedShowLayout_1.default;
var Tab_1 = __importDefault(require("./Tab"));
exports.Tab = Tab_1.default;
var TabbedShowLayoutTabs_1 = __importDefault(require("./TabbedShowLayoutTabs"));
exports.TabbedShowLayoutTabs = TabbedShowLayoutTabs_1.default;
